src/tentaclio/clients/gcs_client.py

```diff
@@ -0,0 +1,160 @@
+                will be used. See storage client docs.
+        """
+        super().__init__(url)
+
+        self._init_creds(project, creds)
+        self.bucket = self.url.hostname or None
+        self.key_name = self.url.path[1:] if self.url.path != "" else ""
+
+        if self.bucket == "gcs":
+            self.bucket = None
+
+    def _connect(self) -> storage.Client:
+        return storage.Client(credentials=self.creds, project=self.project)
+
+    def close(self) -> None:
+        """Close the connection.
+
+        This is fake. GCS doesn't allow to close gcs connections.
+        """
+        # gcs doesn't have close method
+        if self.closed:
+            raise ValueError("Trying to close a closed client")
+        self.closed = True
+
+    # Stream methods:
+
+    @decorators.check_conn
+    def get(
+        self, writer: protocols.ByteWriter, bucket_name: str = None, key_name: str = None
+    ) -> None:
+        """Download the contents from gcs and write them in the provided writer.
+
+        Arguments:
+            :bucket_name: If not provided in the url at construction time.
+            :key_name: If not provided in the url at construction time.
+        """
+        gcs_bucket, gcs_key = self._fetch_bucket_and_key(bucket_name, key_name)
+
+        try:
+            self._get(writer, gcs_bucket, gcs_key)
+        except google_exceptions.NotFound:
+            raise exceptions.GCSError("Unable to fetch the remote file")
+
+    @decorators.check_conn
+    def put(
+        self, reader: protocols.ByteReader, bucket_name: str = None, key_name: str = None
+    ) -> None:
+        """Up the contents to gcs.
+
+        Arguments:
+            :bucket_name: If not provided in the url at construction time.
+            :key_name: If not provided in the url at construction time.
+        """
+        gcs_bucket, gcs_key = self._fetch_bucket_and_key(bucket_name, key_name)
+
+        try:
+            self._put(reader, gcs_bucket, gcs_key)
+        except google_exceptions.NotFound:
+            raise exceptions.GCSError("Unable to fetch the remote file")
+
+    # Helpers:
+
+    def _get_default_creds(self) -> Tuple[credentials.Credentials, Optional[str]]:
+        """Get the default creds.
+
+        Wrapper for the default function so it can be mocked.
+        Unable to mock the default function thus added a wrapper.
+        """
+        return default()
+
+    def _init_creds(
+        self,
+        project: Optional[str] = None,
+        creds: Optional[credentials.Credentials] = None
+    ) -> None:
+        default_creds, default_project = self._get_default_creds()
+        self.creds = creds or default_creds
+        self.project = project or default_project
+
+        if self.project is None:
+            raise exceptions.GCSError("Must have a project set to use gcs.")
+
+        if self.creds is None:
+            raise exceptions.GCSError("Must have credentials to use gcs.")
```

Comment:
I think that `google.auth.default` raises an exception if there are no default creds, so 
a) if the creds are set manually & there are no defaults it will (wrongly) fail
b) the exceptions.GCSError will never get thrown - instead the google client library will throw the google error

In fact, I think that `storage.Client` will call `default` under the hood if the credentials / project is left to default anyway, so you avoid doing the work here in `_init_creds` and leave it all to figure it out in `_connect`. Could leave the `google.auth.DefaultCredentialsError` there, or wrap it in a tentaclio exception.